saleor/graphql/reservation/mutations.py

```diff
@@ -0,0 +1,299 @@
+from datetime import datetime, timedelta
+from typing import TYPE_CHECKING, List, TypedDict
+
+import graphene
+from django.conf import settings
+from django.core.exceptions import ValidationError
+from django.db import transaction
+from django.utils import timezone
+from graphql.error import GraphQLError
+
+from ...core.exceptions import InsufficientStock
+from ...reservation import models
+from ...reservation.error_codes import ReservationErrorCode
+from ...reservation.stock import (
+    get_user_reserved_quantity_bulk,
+    remove_user_reservations,
+)
+from ...shipping.models import ShippingZone
+from ...warehouse.availability import check_stock_quantity
+from ..account.enums import CountryCodeEnum
+from ..channel import ChannelContext
+from ..core.mutations import BaseMutation
+from ..core.types.common import ReservationError
+from ..product.types import ProductVariant
+from .types import RemovedReservation, Reservation
+
+if TYPE_CHECKING:
+    from ...account.models import User
+
+RESERVATION_LENGTH = timedelta(minutes=10)
+RESERVATION_SIZE_LIMIT = settings.MAX_CHECKOUT_LINE_QUANTITY
+RESERVATIONS_TO_REMOVE_LIMIT = 50
+
+
+def _get_reservation_expiration() -> "datetime":
+    return timezone.now() + RESERVATION_LENGTH
+
+
+class ReservationCreateInput(graphene.InputObjectType):
+    country_code = CountryCodeEnum(description="Country code.", required=True)
+    quantity = graphene.Int(
+        required=True, description="The number of items to reserve."
+    )
+    variant_id = graphene.ID(required=True, description="ID of the product variant.")
+
+
+class ReservationCreate(BaseMutation):
```

Comment:
Why the reservation flow is a logic separated from the checkout flow? By adding this we will have two different path of going through the checkout, right?  Why? Shouldn't we include it directly in the checkout flow and allow staff users to control if this feature should be enabled?  This is an additional checkout path to maintain.

In a few weeks, we are going to work on the possibility to extend shipping method by external services, which will need to reserve the stock on the checkout level - as we need to determine the address of "shipFrom" (warehouse address).

Reply:
cc @patrys

Reply:
This will be changed. I've talked to Rafał about how the approach changed after talking to some customers and partners.

Reply:
By saying `will be changed` do you mean the current feature branch or sometime in the future?

Reply:
I mean "in this pull request" 😄